From: Boqun Feng <boqun.feng@gmail.com>
Date: Tue, 23 Nov 2021 15:09:26 +0800
Subject: virt: Fix the detection for Hyper-V VMs

Use product_version instead of product_name in DMI table and the string
"Hyper-V" to avoid misdetection.

Fixes: #21468

Signed-off-by: Boqun Feng <boqun.feng@gmail.com>
---
 src/basic/virt.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/src/basic/virt.c
+++ b/src/basic/virt.c
@@ -148,7 +148,8 @@ static int detect_vm_dmi(void) {
                 "/sys/class/dmi/id/product_name", /* Test this before sys_vendor to detect KVM over QEMU */
                 "/sys/class/dmi/id/sys_vendor",
                 "/sys/class/dmi/id/board_vendor",
-                "/sys/class/dmi/id/bios_vendor"
+                "/sys/class/dmi/id/bios_vendor",
+                "/sys/class/dmi/id/product_version" /* For Hyper-V VMs test */
         };
 
         static const struct {
@@ -166,7 +167,7 @@ static int detect_vm_dmi(void) {
                 { "Parallels",     VIRTUALIZATION_PARALLELS },
                 /* https://wiki.freebsd.org/bhyve */
                 { "BHYVE",         VIRTUALIZATION_BHYVE     },
-                { "Microsoft",     VIRTUALIZATION_MICROSOFT },
+                { "Hyper-V",       VIRTUALIZATION_MICROSOFT },
         };
         unsigned i;
         int r;
